Add public API for dynamic real-time MIDI playback control

nehochupechatat opened the pull request:
This pull request adds support for:
- A public API that would allow the end user to control MIDI playback realtime by muting/unmuting tracks and channels, transposing them and sending pitch bends/progrsm changes/control changes. Very useful for video game developers in particular.
- 'loopstart'/'loopend' markers in MIDI files (case-insensitive). Many MIDIs use this.

derselbst commented on src/midi/fluid_midi.c:
```diff
@@ -1719,8 +1816,8 @@ new_fluid_player(fluid_synth_t *synth)
     fluid_atomic_int_set(&player->seek_ticks, -1);
-    fluid_player_set_playback_callback(player, fluid_synth_handle_midi_event, synth);
-    fluid_player_set_tick_callback(player, NULL, NULL);
+    fluid_player_set_playback_callback(player, fluid_synth_handle_midi_event, player);
```
This effectively passes the `player` object to `fluid_synth_handle_midi_event`, which however expects a `synth` object. This will effectively crash fluidsynth, which our failing CI builds also prove. It also means that your statement to have "thoroughly tested this" is not credible!

derselbst commented on src/midi/fluid_midi.c:
```diff
@@ -1719,8 +1816,8 @@ new_fluid_player(fluid_synth_t *synth)
-    fluid_player_set_playback_callback(player, fluid_synth_handle_midi_event, synth);
-    fluid_player_set_tick_callback(player, NULL, NULL);
+    fluid_player_set_playback_callback(player, fluid_synth_handle_midi_event, player);
+    fluid_player_set_tick_callback(player, fluid_player_handle_loops, player);
```
It should be documented in `fluid_player_set_tick_callback` that loops will no longer work if the user registers his own tick_callback. Which is why this function was invented.

derselbst commented on src/midi/fluid_midi.c:
```diff
@@ -2884,3 +3005,414 @@ fluid_midi_event_length(unsigned char event)
+	
+	/* return if loopstart tick is undefined in MIDI or */
+	/* loopend tick is undefined in MIDI and player hasn't reached song end tick */
+	if(!player->loopstart_tick || (!player->loopend_tick && tick < total_ticks))
```
So this means that if your loopstart is at the very beginning, the very first tick of the MIDI file, your looping logic won't work, wouldn't it?

nehochupechatat commented on src/midi/fluid_midi.c:
```diff
@@ -1617,23 +1661,63 @@ fluid_track_send_events(fluid_track_t *track,
+			if ((event->type == NOTE_ON || event->type == NOTE_OFF) && player->transposed_ch[event->channel] && player->transpose != 0)
+			{
+				/* Limit transposition amount to avoid sending invalid MIDI note numbers */
+				if(player->transpose < -24) {player->transpose = -24;} else if (player->transpose > 24) {player->transpose = 24;}
```
Because there are MIDI files that use basses with note range lower than 36 (octave 24-36).

nehochupechatat commented:
While I have fixed most of the glitches there's just one problem.

Even if I remove the "check if loopstart tick is zero" function, my testing with fluidsynth.c still shows that if there are no markers, it will still just loop with the "playlist" mechanism, resetting everything in the process.

The track muting is now determined through player->ntracks, but that's not very convinient because it only gets it upon starting. And since it's asynchronous, you can't just call play and then get the track number... I was perhaps thinking of moving the ntrack declaration somewhere earlier?

Yes, I changed the APIs to your suggestions, but I kept the midi Ctrl change enums so game developers could send custom Control Change events. Will commit soon

derselbst commented:
I'm closing this now. For the future, pls. consult the [Contributing guideline](https://github.com/FluidSynth/fluidsynth/blob/master/CONTRIBUTING.md) of a project, before actually contributing anything. I admit, our guidelines are a bit out of date. Yet, it would have prevented at least 40% of code-related problems in this PR, if you had read it.
